src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -1449,28 +1456,49 @@ private ValueTask WriteToStreamAsync(ReadOnlyMemory<byte> source, bool async)
             }
         }
 
-        private bool TryReadNextLine(out ReadOnlySpan<byte> line)
+        private bool TryReadNextChunkedLine(bool readingHeader, out ReadOnlySpan<byte> line)
         {
+            int maxByteLength = readingHeader ? _allowedReadLineBytes : MaxChunkBytesAllowed;
             var buffer = new ReadOnlySpan<byte>(_readBuffer, _readOffset, _readLength - _readOffset);
-            int length = buffer.IndexOf((byte)'\n');
-            if (length < 0)
+
+            int lfIdx = buffer.IndexOf((byte)'\n');
+            if (lfIdx < 0)
             {
-                if (_allowedReadLineBytes < buffer.Length)
+                if (buffer.Length < maxByteLength)
                 {
-                    throw new HttpRequestException(SR.Format(SR.net_http_response_headers_exceeded_length, _pool.Settings._maxResponseHeadersLength * 1024L));
+                    line = default;
+                    return false;
                 }
+            }
+            else
+            {
+                int bytesConsumed = lfIdx + 1;
+                int maxBytesRemaining = maxByteLength - bytesConsumed;
+                if (maxBytesRemaining >= 0)
+                {
+                    _readOffset += bytesConsumed;
 
-                line = default;
-                return false;
+                    if (readingHeader)
+                    {
+                        _allowedReadLineBytes = maxBytesRemaining;
+                    }
+
+                    int crOrLfIdx = lfIdx - 1;
+                    if (crOrLfIdx < 0 || buffer[crOrLfIdx] != '\r')
```

Comment:
```suggestion
                    if ((uint)crOrLfIdx >= (uint)buffer.Length || buffer[crOrLfIdx] != '\r')
```
to avoid the bounds check?